Add `-UnixTimeSeconds` to `Get-Date` to allow Unix time input

aetos382 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Add the `-UnixTimeSeconds` parameter to the `Get-Date` command.
This allows the time in seconds elapsed since 1970-01-01T00:00:00Z (aka Unix Time).

PR #12179 has already been submitted, but my proposal is more like the specification proposed in #11719.
This proposal also supports negative values (meaning time before the Unix epoch), therefore improving compatibility with the [DateTimeOffset](https://docs.microsoft.com/en-us/dotnet/api/system.datetimeoffset) class and the Linux `date` utility.

The following is an example of its use.

```powershell
> Get-Date -UnixTimeSeconds 1593696159

Thursday, July 2, 2020 1:22:39 PM

> Get-Date -UnixTimeSeconds -1

Wednesday, December 31, 1969 11:59:59 PM

```

The parameter has `-UnixTime` alias, because 'Unix Time' generally means a value expressed in seconds.
Also, consider the possibility of adding a parameter in the future that takes values expressed in milliseconds.

The parameter is named `UnixTimeSeconds`, not `FromUnixTime`.
This is for consistency with the `Date` parameter (it's not `FromDate`).

## PR Context

Close #12976

## Compatibility Notes
This proposal is basically compatible with PowerShell 7.0.2, but not compatible with 7.1.0-preview.4, including #12179.

It also includes a parameter set name changes. It's classified as Non-Public surface in the [guideline](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-4-clearly-non-public). 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/6263
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

aetos382 committed:
add ParameterSetNames

aetos382 committed:
use ParameterSetNames

aetos382 committed:
change FromUnixTime to UnixTimeSeconds

aetos382 committed:
validate range of UnixTimeSeconds

aetos382 committed:
add property body of Format and UFormat

aetos382 committed:
make Format and UFormat are exclusive.

Since it was difficult to implement exclusivity between Date and UnixTimeSeconds, and Format and UFormat using parameter set, I checked Format and UFormat at runtime.

aetos382 committed:
use indicator fields

aetos382 committed:
add tests

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs:
```diff
@@ -212,26 +235,73 @@ public int Millisecond
+            set
+            {
+                _format = value;
+                _formatSpecified = true;
```
We have no need to add the field - use `this.MyInvocation.BoundParameters[]` to check if a parameter is present.

iSazonov replied:
Perhaps we could create parameter sets to resolve this.

aetos382 replied:
For Format and UFormat, I made it parameter set (thanks to @mklement0).

For Date and UnixTimeSeconds, the other parameters are not checked by BoundParameters, and use a discernment field, so I'm following them.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs:
```diff
@@ -243,14 +313,11 @@ protected override void ProcessRecord()
+            }
+            else if (_unixTimeSecondsSpecified)
+            {
+                dateToUse = DateTimeOffset.FromUnixTimeSeconds(UnixTimeSeconds).LocalDateTime;
```
Do you consider `-AsUTC ` too?

mklement0 replied:
As an aside, @iSazonov: Do you know why `-AsUTC` cannot be combined with `-UFormat`? Was that intentional?

iSazonov replied:
I think it can be in UFormat set too.

aetos382 replied:
The AsUTC parameter can be combined with all other parameters.

iSazonov replied:
@aetos382 Please add AsUTC support.

aetos382 replied:
It's supported here (it's the same code path as using -Date -AsUTC).

```cs
if (AsUTC)
{
    dateToUse = dateToUse.ToUniversalTime();
}
```

iSazonov replied:
I mean why do you interpret `UnixTimeSeconds` as local Datetime here?

aetos382 replied:
Because `dateToUse` is local time (it's not `DateTime.UtcNow`).

```cs
protected override void ProcessRecord()
{
    DateTime dateToUse = DateTime.Now;
```

This command consistently handles the local time unless an `-AsUTC` parameter is set.

If the `-AsUTC` parameter is set, it will be handled by the code cited in the above comment, and therefore it has not been considered here.

mklement0 replied:
I agree that outputting a local date by default makes sense with `-FromUnixTime` too, but to provide some background:

`Get-Date` _defaults to_ a local `[datetime]` instance (one whose `.Kind` property value is `Local`),  (`(Get-Date).Kind`) and you also get such an instance if you modify the current date with parameters such as `-Year`
 (e.g., `(Get-Date -Year 2019).Kind`).

By contrast, if you pass an argument to `-Date`:

* If you pass a _preexisting_ `[datetime]` instance, its `.Kind` property is retained (e.g.
`(Get-Date ([datetime]::UtcNow)).Kind` (`Utc`)

* If you pass a _string_ that is implicitly _converted to_ `[datetime]` _during parameter binding_ (that is, this logic is outside of `Get-Date`):
   * If that string _doesn't_ contain a time-zone offset, the `.Kind` property is `Unspecified`; e.g.,
 `(Get-Date 2020-01-01).Kind` (`Unspecified`) 
    * If it _does_ contain a time-zone offset (whatever it is, even if it indicates UTC), the `.Kind` property is `Local`; that is, _conversion to local time_ takes place; e.g., 
`(Get-Date 2020-01-01Z).Kind` (converted from UTC to `Local`).

With respect to string representations and `.ToUniversalTime()`, `Unspecified` instances behave like `Local` ones.
However, calling `.ToLocalTime()` on an `Unspecified` instance interprets it as _UTC_ time.

iSazonov replied:
Ok, if we accept the default we should add a test and document this.

aetos382 replied:
Hmm...

Would it be better to have `-Kind <DateTimeKind>` parameter instead of an `-AsUTC`?

iSazonov replied:
I guess the DateTimeKind was added to mitigate Datetime type limitations (today we could use DateTimeOffset). I do not think that we should use DateTimeKind broadly. At least we could discuss this in new issue.

mklement0 replied:
In fact, there already _is_ a proposal to implement this: #11731 - and I do think it makes perfect sense.

Whether or not we implement it as part of this PR, I think it's important to also include it in v7.1

aetos382 replied:
I think a parameter like `-AsTimeZone PDT` would be useful.

Then I would also want features like `-FromTimeZone PDT -ToTimeZone JST`.

But adding those features to `Get-Date` would make the command overly complicated.
For example, I think we should add another command, such as `Convert-Time`.

Should there be a place for a comprehensive discussion on how `Get-Date` should handle timezone and offset-related parameters (including `-AsUTC`)?

iSazonov replied:
Let's continue in  #11731.

mklement0 replied:
@iSazonov: See my comment at https://github.com/PowerShell/PowerShell/issues/11731#issuecomment-659279816.

I think that potential future `[datetimeoffset]` and time-zone (conversion) support should be discussed in a _new_ issue, which I encourage you, @aetos382, to create (as for the name `Convert-Time`: for consistency it's better to stick with the word "date").

iSazonov replied:
@aetos382 You can open new issue to discuss Convert-Time/Date/Datetime cmdlet.

iSazonov commented:
@jackdcasey @mklement0 Please review.

mklement0 commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs:
```diff
@@ -212,26 +235,73 @@ public int Millisecond
+        /// </summary>
+        protected override void BeginProcessing()
+        {
+            if (_formatSpecified && _uFormatSpecified)
```
I think we should use parameter sets instead so as to let the parameter binder enforce mutually exclusive parameter; while their number can get out of hand, here we only need 4: `DateTime` (with  optional `-Format`), `DateTimeUFormat`, `UnixTime`  (with  optional `-Format`), and `UnixTimeUFormat`

aetos382 committed:
Using a ParameterSet to control parameter exclusivity

aetos382 committed:
Remove unnecessary code

aetos382 committed:
Set ParameterSetName for OutputType.

If UFormat is specified, DateTime is never returned.

aetos382 committed:
Fix documentation comment

To pass CodeFactor's static check, the value of the summary element now begins with an alphabetic character.

aetos382 committed:
Fix test

By changing the validation attribute of the UFormat parameter from ValidateNotNullOrEmpty to Mandatory

aetos382 committed:
Remove test

This test is not necessary because ParameterSet is used to ensure that Format and UFormat are not used together, and is not determined by an if statement.

daxian-dbw commented:
@TravisEz13 and @iSazonov, since both of you reviewed and approved #12179, can you please also review this PR?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs:
```diff
@@ -544,6 +567,17 @@ private string UFormatDateString(DateTime dateTime)
 
         #endregion
+
+        private static class ParameterSetNames
```
Our common patter is to use consts without extra static class.

aetos382 replied:
Did it.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1:
```diff
@@ -193,13 +193,15 @@ Describe "Get-Date" -Tags "CI" {
 
-    It "-FromUnixTime works" {
+    It "-UnixTimeSeconds works" {
+        # Test conversion of arbitrary date in Unix time: 2020-01-01T00:00:00.000Z
```
It is preferred to use `TestCases`.

aetos382 replied:
I did so.

iSazonov commented:
@aetos382 Please address my comments.

aetos382 committed:
Use testcase

aetos382 committed:
Declare ParameterSetNames directly.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs:
```diff
@@ -40,14 +41,38 @@ public DateTime Date
+        private DateTime _date;
+        private bool _dateSpecified;
+
+        // DateTimeOffset.MinValue.ToUnixTimeSeconds()
```
Looks like commented code. Please reword like:
```suggestion
        // The const comes from DateTimeOffset.MinValue.ToUnixTimeSeconds()
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs:
```diff
@@ -40,14 +41,38 @@ public DateTime Date
+        // DateTimeOffset.MinValue.ToUnixTimeSeconds()
+        private const long MinimumUnixTimeSecond = -62135596800;
+
+        // DateTimeOffset.MaxValue.ToUnixTimeSeconds()
```
The same.
```suggestion
        // The const comes from DateTimeOffset.MaxValue.ToUnixTimeSeconds()
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs:
```diff
@@ -544,6 +567,11 @@ private string UFormatDateString(DateTime dateTime)
 
         #endregion
+
+        private const string DateAndFormat = "DateAndFormat";
```
It is parameter set names. So I'd add this in the consts:
```suggestion
        private const string DateAndFormatParameterSet = "DateAndFormat";
```

aetos382 replied:
I think that if the parameter set name constants have the suffix `ParameterSet`, it would be better to make them members of the static inner class, as before.

iSazonov replied:
It is our common pattern. If you see a const with the suffix in code you understand that it is a parameter set name.

aetos382 committed:
Update src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs

Co-authored-by: Ilya <darpa@yandex.ru>

aetos382 committed:
Update src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs

Co-authored-by: Ilya <darpa@yandex.ru>

aetos382 committed:
add 'ParameterSet' suffix to parameter set names

daxian-dbw commented:
@iSazonov Thanks for the review!
@TravisEz13 can you please also take a look?

daxian-dbw commented:
@aetos382 and @iSazonov, I notice the PR has a breaking change comparing to the existing 7.1 preview versions -- `FromUnixTime` is removed. Is that breaking change intentional? (just want to confirm).

aetos382 commented:
@daxian-dbw 
> Is that breaking change intentional? (just want to confirm).

Yes. I am aware from the beginning that this PR is not compatible with the 7.1.0 preview.

iSazonov commented:
@daxian-dbw  Yes, new approach looks more good and it better to replace previous commit before RC and release. 
I think we have no need use "Breaking-Change" label because it is used only for real breaking changes with released versions. Otherwise docs and release notes will confuse users.

daxian-dbw commented:
Thank you both for the clarification.
I will leave the "breaking change" label on for now and marked this PR to be reviewed by maintainers, so we can conclude on how to deal with breaking changes that happen within the preview releases.

iSazonov commented:
@aetos382 Thanks for your contribution!

aetos382 commented:
@iSazonov @daxian-dbw @mklement0 @SteveL-MSFT Thank you all!

scotthardwick commented:
@aetos382 Thank you so much for your work on this!  I wrote https://github.com/PowerShell/PowerShell/issues/11719 and am glad to see this come to completion. It will be very useful.  Also look forward to your work on Timezones in https://github.com/PowerShell/PowerShell/issues/13312
Finally, if you get a chance, please feel free to look at my issue https://github.com/PowerShell/PowerShell/issues/11714 and make any suggestions too. It is also tied up with Unix Date/Times.
